src/Components/Components/src/Reflection/ComponentResolver.cs

```diff
@@ -6,28 +6,45 @@
 using System.Linq;
 using System.Reflection;
 
-namespace Microsoft.AspNetCore.Components
+namespace Microsoft.AspNetCore.Components.Reflection
 {
     /// <summary>
     /// Resolves components for an application.
     /// </summary>
-    internal static class ComponentResolver
+    public sealed class ComponentResolver
     {
-        /// <summary>
-        /// Lists all the types
-        /// </summary>
-        /// <param name="appAssembly"></param>
-        /// <returns></returns>
-        public static IEnumerable<Type> ResolveComponents(Assembly appAssembly)
+        private Assembly _cachedAssembly;
+        private IReadOnlyList<Type> _resolvedComponents;
+
+        public IReadOnlyList<Type> ResolveComponents(Assembly assembly)
+        {
+            if (_cachedAssembly == assembly)
+            {
+                return _resolvedComponents;
+            }
+
+            var resolvedComponents = ResolveComponentsCore(assembly);
+
+            if (_cachedAssembly is null)
+            {
+                _resolvedComponents = resolvedComponents;
+                _cachedAssembly = assembly;
+            }
+
+            return _resolvedComponents;
```

Comment:
We do plan on supporting multipole roots, so we need some kind of way to allow that here.

Reply:
Should probably be concdict